server/worker/tasks.py

```diff
@@ -128,28 +124,25 @@ def run_task(task: BackgroundTask, db_session=db_session):
             sentry_sdk.capture_exception(error)
 
 
-def run_new_tasks(db_session=db_session):
-    # Cleanup any tasks that failed to finish processing last time the worker was run
-    stuck_tasks = (
+def claim_next_task(db_session=db_session) -> Optional[BackgroundTask]:
+    task: Optional[BackgroundTask] = (
         db_session.query(BackgroundTask)
-        .filter(BackgroundTask.started_at.isnot(None))
-        .filter(BackgroundTask.completed_at.is_(None))
-        .all()
+        .filter_by(started_at=None)
+        .order_by(BackgroundTask.created_at)
+        .with_for_update()
+        .limit(1)
+        .one_or_none()
     )
-    for task in stuck_tasks:
-        task.started_at = None
-        logger.info(f"TASK_RESET {task_log_data(task)}")
+    if task:
+        task.started_at = datetime.now(timezone.utc)
```

Comment:
is it worth recording maybe a process ID in addition to the time at which it's started, in case we need to do cleanup later?

Reply:
Can you say a little more about how you envision the process ID enabling cleanup? What sort of situations would this help with?

Reply:
possibly a situation where a process dies and doesn't get a proper chance to clean up, then there's a trace of which process it was that grabbed it. Partly I'm nervous about the state of who owns a task being only stored in RAM.

Reply:
I see. If we could differentiate the currently running workers from past workers, then I see how this could be useful. I bet there's some ID assigned to each dyno that we could access.

That being said, I think it's also technically solvable if we know how many workers are currently running. E.g. if there are 2 running workers and 3 active tasks, the active task with the oldest start time is not actually being run by a worker and should be cleaned up.

I'm not opposed to the process ID though, I agree it could come in handy hypothetically.